File: Year-2023/LeetCode-May-Challenge/KthLargestElementinaStream.cpp
#include<bits/stdc++.h>
using namespace std;
class KthLargest {
private:
    int kth;
    priority_queue<int, vector<int>, greater<int>> pq;
public:
    KthLargest(int k, vector<int>& nums) {
        kth = k;
        for(auto x: nums) add(x);
    }

    int add(int val) {
        if(pq.size() < kth) pq.push(val);
        else if(pq.top() < val) pq.pop(), pq.push(val);
        return pq.top();g
    }
};

int main(){
    vector<int> nums = {4,5,8,2};
    int k = 3;
    KthLargest *s = new KthLargest(k, nums);
    cout<<s->add(3)<<endl;
    cout<<s->add(5)<<endl;
    cout<<s->add(10)<<endl;
    cout<<s->add(9)<<endl;
    cout<<s->add(4)<<endl;
    return 0;
}

File: Year-2023/LeetCode-May-Challenge/KthLargestElementinaStream.java
import java.util.*;

class KthLargest {
    private int kth;
    private PriorityQueue<Integer> pq = new PriorityQueue<>();

    public KthLargest(int k, int[] nums) {
        kth = k;
        for (int x : nums)
            add(x);
    }

    public int add(int val) {
        if (pq.size() < kth)
            pq.add(val);
        else if (pq.peek() < val) {
            pq.remove();
            pq.add(val);
        }
        return pq.peek();
    }
}

public class KthLargestElementinaStream {
    public static void main(String[] args) {
        int[] nums = { 4, 5, 8, 2 };
        int k = 3;
        KthLargest s = new KthLargest(k, nums);
        System.out.println(s.add(3));
        System.out.println(s.add(5));
        System.out.println(s.add(10));
        System.out.println(s.add(9));
        System.out.println(s.add(4));
    }
}
